### PR TITLE
feat: computation of shared measurement in CKF

### PR DESCRIPTION
**Pull request by CarloVarni**
Computation of shared measurements in CKF. After track reconstruction, the code runs on all reconstructed tracks and takes into account the measurements used in each track by accessing the measurement index contained in the source link. If the same measurement is used by more than one track, it is flagged as a SharedHit.

The measurement will have both the Measurement and the Shared flag, so no change in measurement-related variables is expected. 

In attachment the number of Shared Measurement distribution as a function of eta and pt. 
[shared_vs_eta.pdf](https://github.com/acts-project/acts/files/6879934/shared_vs_eta.pdf)
[shared_vs_pt.pdf](https://github.com/acts-project/acts/files/6879935/shared_vs_pt.pdf)

Here is also the number of shared hits for tracks: [nSharedHits_raw.pdf](https://github.com/acts-project/acts/files/6879937/nSharedHits_raw.pdf)
Concerning the color legend: 
- black=good track
- red=duplicate track
- green=fake track
 
the labelling is taken from [CsvMultiTrajectoryWriter.cpp](https://github.com/acts-project/acts/blob/main/Examples/Io/Csv/src/CsvMultiTrajectoryWriter.cpp#L139)

The high number of shared measurements is a result of the CKF reconstructing several tracks from the same truth particles. These reconstructed tracks shared almost all the measurements. 
For instance this particular case in which three different tracks are reconstructed 
[recoParticlesFor_78019200.pdf](https://github.com/acts-project/acts/files/6879949/recoParticlesFor_78019200.pdf):
```
[Measurement indexes] [track type]
[6851, 10338, 16437, 16430, 25254, 25320, 31900, 32675, 33246, 33979, 40070] [duplicate]
[6851, 10338, 16435, 16430, 17039, 25254, 25320, 31900, 32675, 33246, 33979, 40070] [duplicate]
[6851, 10338, 16435, 16429, 17039, 25254, 25320, 31900, 32675, 33246, 33979, 40070] [good]
```

Pinging @XiaocongAi @paulgessinger @asalzburger

### CODE REVIEW

**Comment by CarloVarni** [comment]
Here is the shared hits distribution if we recompute that quantity taking into account the truth particle information (i.e. measurements used **only** by reconstructed tracks from the same truth particle are not labelled as shared): [nSharedHits_recomputed.pdf](https://github.com/acts-project/acts/files/6880097/nSharedHits_recomputed.pdf)

**Comment by paulgessinger** [comment]
Is this WIP or ready to review?

**Comment by CarloVarni** [comment]
> Is this WIP or ready to review?

Hi @paulgessinger , I think this PR is ready to review.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1316]: This data structure probably does a ton of heap allocations. A few ideas would be to:

1. Try to use a *small vector* if possible to save some allocations
2. Try not to use a map but rather a sufficiently sized vector (since the keys should be an integer sequence **I think**
3. Use a pool allocator for both map and vectors.

Comment (XiaocongAi) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1316]: I think it might be better to not check about the shared hits during the track finding. The shared hits info done in this way is not accurate anyway since one CKF invocation only deals with one seed. But it's very likely that the shared hits happen between tracks found via multiple different seeds. 
Hence, a better solution might be that we sort out the shared hits after the all the tracks have been found if necessary. This is more straightforward and accurate.

Comment (CarloVarni) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1316]: Hi @XiaocongAi, I see however that the `findtrack` method (as called in [TrackFindingAlgorithm.cpp#L68](https://github.com/acts-project/acts/blob/main/Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp#L68)) uses all the initial parameters and measurements. I suppose this is taking into consideration the entire events, right? 

Inside the CombinatorialKalmanFilter.hpp file there is indeed a loop over all the seeds. That is on line [CombinatorialKalmanFilter.hpp#L1265](https://github.com/CarloVarni/acts/blob/main/Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp#L1265 ). However my changes would be after that loop. Shouldn't we have all the tracks at that point, from all the seeds?

Is my understanding right?

We cal always move the computation somewhere else I think to make the code more straigthforward. Would we have all tracks in [TrackFindingAlgorithm.cpp#L89](https://github.com/acts-project/acts/blob/main/Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp#L89)? Or a different (new) algorithm can be added to the sequence.

Comment (XiaocongAi) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1316]: Also, as @paulgessinger suggested, we might use a vector for the `recordedHits` such that the element `i` in the vector would correspond to the info for `hitIndex i` .

**Review by XiaocongAi — COMMENTED** [review]
Hi @CarloVarni , I have some further comments, which should be easy to address. I will approve once they are addressed.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp
Comment (CarloVarni) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp]: Hi @XiaocongAi, I have moved the shared hits computation inside `Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp` in a ad-hoc method.

Also, as @paulgessinger suggested, I have modified the algorithms in order not use `std::unordered_map` but to use two `std::vector<int>` instead.

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp
Comment (XiaocongAi) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:71]:
```suggestion
  if (m_cfg.computeSharedHits){
```

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp
Comment (XiaocongAi) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:73]:
```suggestion
}
```

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp
Comment (XiaocongAi) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:117]:
```suggestion
    if (not results.at(iresult).ok()){
```

**Review by XiaocongAi — APPROVED** [review]
Hi @CarloVarni , thank you for the changes. It looks good to me.

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/TrackFindingAlgorithm.hpp:114]: Maybe `size_t` and `std::numeric_limits<size_t>::max()` instead of `-1`?

**Comment by asalzburger** [comment]
Updating this ...

**Review by XiaocongAi — APPROVED** [review]

**Comment by paulgessinger** [comment]
I think we'll have to wait for one more round of CI before merging.

**Pull request merged**
